### PR TITLE
cpu/esp32: fix compilation issues with GCC 12.2

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides the changes in `cpu/esp32` and `cpu/esp_common` to fix the compilation issues with GCC v12.2.  It is required as the first step in the preparation of the upgrade to ESP-IDF version 5.1.

**Please note**: Insead of fixing the ESP-IDF 4.4 code itself by a big bunch of patches to fix the compilation problems with GCC v12.2, it temporarily disables some warnings. The reason is that the ESP-IDF 5.1 requires GCC v12.2 and should be fixed for this compiler version by the vendor.

### Testing procedure

Green CI

The change were already tested with all ESP-specific modules like `esp_now`, `esp_wifi`, `esp_spi`  and `esp_ble` for all supported ESP platforms.

### Issues/PRs references

Prerequisite for https://github.com/RIOT-OS/riotdocker/pull/227
Fixes issue #19421

### CODE REVIEW

**Comment by gschorcht** [comment]
bors try

**Comment by gschorcht** [comment]
bors cancel

**Comment by gschorcht** [comment]
bors try

**Comment by gschorcht** [comment]
bors try

**Comment by gschorcht** [comment]
Finally it compiled and worked for me locally with gcc 12.2 and still compiled in Murdock with gcc 8.4.0. So it seems to be ready.

**Review by maribu — APPROVED** [review]
Looks good to me. One remark inline. If you say it is worth to go for `-misa-spec-2.2` instead, please squash that right in. Otherwise this can go in as is.

**Comment by maribu** [comment]
The static tests are not yet happy. Please squash at will

**Comment by gschorcht** [comment]
Ah sorry, I pushed accidentally the branch with PR #19452 included that caused the static test error. I removed this commit.

**Comment by gschorcht** [comment]
rebased

**Comment by gschorcht** [comment]
@maribu I have changed the evaluation whether option `-misa-spec=2.2` has to be used to the more general approach used in `makefiles/arch/riscv.inc.mk`. Does your approval still applies?

**Comment by maribu** [comment]
bors merge

**Comment by maribu** [comment]
I do still approve, but the check-commits hook insists on squashing

**Comment by gschorcht** [comment]
> I do still approve, but the check-commits hook insists on squashing

I just wanted to keep the latest changes visible.

**Comment by gschorcht** [comment]
I compared my changes with yours in PR #18719. Some of them are the same but a number of changes differ. A number of differences are caused by the cleanup in PR #19426, e.g, `cpu/esp32/periph/adc.c` vs. `cpu/eps32/esp-id-api/adc.c`. Some of your changes seem to make more sense, e.g., `cpu/esp32/esp-idf/esp_idf_support.c`.

Should we compare and discuss the differences again? For example, my `cpu/esp32/Makefile.include` includes the changes needed for ESP32-C3 while yours include `CFLAGS += -D__DYNAMIC_REENT__`.

Your PR also includes changes for ESP8266 I would like to get in to be able to upgrade to a newer GCC version.

**Comment by maribu** [comment]
Maybe the easiest would be if I just close https://github.com/RIOT-OS/RIOT/pull/18719 in favor of this PR - I haven't really looked actively into or updated the PR for some time anyway. If there is something useful to scavenge from (e.g. regarding ESP8266), you are welcome to do so.

File: cpu/esp32/esp-idf/esp_idf_support.c [hunk 1]
```diff
@@ -111,7 +111,7 @@ void IRAM_ATTR esp_log_writev(esp_log_level_t level,
      * We use the log level set for the given tag instead of using
      * the given log level.
      */
-    esp_log_level_t act_level = LOG_DEBUG;
+    esp_log_level_t act_level = (esp_log_level_t)LOG_DEBUG;
```
Comment (gschorcht) [inline cpu/esp32/esp-idf/esp_idf_support.c:114]: @maribu I'm wondering what the cleaner approach would be here. Casting to RIOT's `LOG_DEBUG` to `esp_log_level_t` or using `ESP_LOG_DEBUG` as in PR #18532. the enumeration `esp_log_level_t` is compatible with RIOT's log levels.

Comment (maribu) [reply cpu/esp32/esp-idf/esp_idf_support.c:114]: I think it is more a matter of personal taste an either is fine

**Comment by benpicco** [comment]
bors merge

**Comment by gschorcht** [comment]
Thanks for the try to merge it. I had to squash it before.

**Comment by gschorcht** [comment]
borse merge

**Comment by maribu** [comment]
Forgot to push the squash commits?

**Comment by gschorcht** [comment]
Indeed.

**Comment by benpicco** [comment]
bors merge

**Comment by gschorcht** [comment]
@maribu @benpicco Thanks for reviewing and merging.

@kaspar030 Next stept would be to try to update Murdock https://github.com/RIOT-OS/riotdocker/pull/227

**Pull request merged**